assets/js/app.js

```diff
@@ -0,0 +1,277 @@
+        //creates the basic dom
+        const loadingDiv = document.getElementById("loadingDiv")
+        loadingDiv.classList.add(`visible`)
+        const AllTheData = await getAllCoins()
+        loadingDiv.classList.remove(`visible`)
+
+        const first100 = AllTheData.slice(0, 100)
+        const HTML = first100.map(coin =>
+            `
+                <div class="card" style="width: 18rem;">
+                   <div class="card-body">
+                        <div class="form-check form-switch">
+                            <input class="form-check-input" type="checkbox" role="switch" id="switch${coin.id}" style= "position: absolute; right: 10px; font-size: xxx-large;">
+                        </div>
+                        <h5 class="card-title">${coin.name}</h5>
+                        <p class="card-text" >${coin.id}</p>
+                        <button type="button" class="btn btn-lg btn-primary" data-bs-toggle="popover" id="${coin.id}" data-bs-content=""> More Info </button>
+                    </div>
+                </div>
+            `)
+            .reduce((cum, cur) => cum + cur, ``)
+        document.getElementById("coinsTabPresentation").innerHTML = HTML
+
+        function allTheFunctionsForCoinPage() {
+            //make popover for all the buttons
+            const popoverTriggerList = document.querySelectorAll('[data-bs-toggle="popover"]');
+            [...popoverTriggerList].map(popoverTrigger => new bootstrap.Popover(popoverTrigger))
+
+            //load 5 coins from memory
+            const loadMemory = JSON.parse(localStorage.getItem(`5coinsInStorage`))
+            loadMemory.forEach(coin => { if (document.getElementById(coin)) document.getElementById(coin).checked = true })
+
+            //the popover button
+            const buttons = document.querySelectorAll("div > button")
+            buttons.forEach(function (button) {
+                button.addEventListener('click', async function () {
+
+                    //save the data to cache key + storage
+                    const coinsCache = JSON.parse(localStorage.getItem(`coinsCache`))
+                    const loadingDiv = document.getElementById("loadingDiv")
+                    const index = coinsCache.findIndex(coin => coin.id === this.id)
+                    let coinData = ``
+                    if (index !== -1) coinData = coinsCache[index]
+                    else {
+                        loadingDiv.classList.add(`visible`)
+                        coinData = await getSingleCoin(this.id)
+                        loadingDiv.classList.remove(`visible`)
+                    }
+
+                    if (!coinsCache.some(coin => coin.id === coinData.id)) coinsCache.push(coinData)
+                    localStorage.setItem(`coinsCache`, JSON.stringify(coinsCache))
+
+                    //after 2 min, deletes from the cache (so in the next call, get the most updated data)
```

Comment:
different approach to cache purging, nice idea (not neccessarily efficient, think why)